File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/EnergyStorageImpl.java
package dev.kineticcat.hexportation.forge.transfer;

import dev.kineticcat.hexportation.api.transfer.EnergyStorage;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraftforge.common.capabilities.ForgeCapabilities;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.energy.IEnergyStorage;

/**
 * Forge implementation that delegates to Forge's IEnergyStorage capability system.
 * Provides identical API to Team Reborn Energy through wrapper pattern.
 */
public class EnergyStorageImpl {
    
    /**
     * Singleton instance that implements our SidedStorage interface
     */
    public static final EnergyStorage.SidedStorage INSTANCE = new EnergyStorage.SidedStorage() {
        @Override
        public EnergyStorage find(ServerLevel level, BlockPos pos, Direction direction) {
            var blockEntity = level.getBlockEntity(pos);
            System.out.println("[EnergyStorage] Checking pos=" + pos + " direction=" + direction + " blockEntity=" + (blockEntity != null ? blockEntity.getClass().getSimpleName() : "null"));
            
            if (blockEntity == null) {
                System.out.println("[EnergyStorage] No blockEntity found, returning null");
                return null;
            }
            
            LazyOptional<IEnergyStorage> capability = blockEntity.getCapability(
                ForgeCapabilities.ENERGY, direction);
            
            boolean hasCapability = capability.isPresent();
            System.out.println("[EnergyStorage] Energy capability present: " + hasCapability);
            
            if (hasCapability) {
                IEnergyStorage handler = capability.orElse(null);
                System.out.println("[EnergyStorage] Found IEnergyStorage: " + handler.getClass().getSimpleName() + " with " + handler.getEnergyStored() + "/" + handler.getMaxEnergyStored() + " RF");
                return capability.map(ForgeEnergyWrapper::new).orElse(null);
            } else {
                System.out.println("[EnergyStorage] No energy capability, returning null");
                return null;
            }
        }
    };
    
    public static long getAmount(Object energyStorage) {
        if (energyStorage instanceof ForgeEnergyWrapper wrapper) {
            return wrapper.getAmount();
        }
        if (energyStorage instanceof IEnergyStorage storage) {
            return storage.getEnergyStored();
        }
        return 0;
    }
    
    public static long getCapacity(Object energyStorage) {
        if (energyStorage instanceof ForgeEnergyWrapper wrapper) {
            return wrapper.getCapacity();
        }
        if (energyStorage instanceof IEnergyStorage storage) {
            return storage.getMaxEnergyStored();
        }
        return 0;
    }
    
    /**
     * Wrapper class that extends our EnergyStorage abstract class using Forge's IEnergyStorage.
     */
    public static class ForgeEnergyWrapper extends EnergyStorage {
        private final IEnergyStorage forgeStorage;
        
        public ForgeEnergyWrapper(IEnergyStorage forgeStorage) {
            this.forgeStorage = forgeStorage;
        }
        
        @Override
        public long getAmount() {
            return forgeStorage.getEnergyStored();
        }
        
        @Override
        public long getCapacity() {
            return forgeStorage.getMaxEnergyStored();
        }
        
        public long simulateExtract(long maxExtract) {
            return forgeStorage.extractEnergy((int) Math.min(maxExtract, Integer.MAX_VALUE), true);
        }
        
        public long simulateInsert(long maxInsert) {
            return forgeStorage.receiveEnergy((int) Math.min(maxInsert, Integer.MAX_VALUE), true);
        }
        
        public long extract(long maxExtract, boolean simulate) {
            return forgeStorage.extractEnergy((int) Math.min(maxExtract, Integer.MAX_VALUE), simulate);
        }
        
        public long insert(long maxInsert, boolean simulate) {
            return forgeStorage.receiveEnergy((int) Math.min(maxInsert, Integer.MAX_VALUE), simulate);
        }
        
        public IEnergyStorage getForgeStorage() {
            return forgeStorage;
        }
    }
}

File: forge/src/main/java/dev/kineticcat/hexportation/forge/transfer/EnergyStorageImpl.kt
package dev.kineticcat.hexportation.forge.transfer

import dev.kineticcat.hexportation.api.transfer.EnergyStorage
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraftforge.common.capabilities.ForgeCapabilities
import net.minecraftforge.energy.IEnergyStorage
import kotlin.math.min

/**
 * Forge implementation that delegates to Forge's IEnergyStorage capability system.
 * Provides identical API to Team Reborn Energy through wrapper pattern.
 */
object EnergyStorageImpl {
    
    /**
     * Singleton instance that implements our SidedStorage interface
     */
    @JvmField
    val INSTANCE = object : EnergyStorage.SidedStorage() {
        override fun find(level: ServerLevel, pos: BlockPos, direction: Direction): EnergyStorage? {
            val blockEntity = level.getBlockEntity(pos) ?: return null
            val capability = blockEntity.getCapability(ForgeCapabilities.ENERGY, direction)
            return capability.map(::ForgeEnergyWrapper).orElse(null)
        }
    }
    
    @JvmStatic
    fun getAmount(energyStorage: Any): Long = when (energyStorage) {
        is ForgeEnergyWrapper -> energyStorage.amount
        is IEnergyStorage -> energyStorage.energyStored.toLong()
        else -> 0
    }
    
    @JvmStatic
    fun getCapacity(energyStorage: Any): Long = when (energyStorage) {
        is ForgeEnergyWrapper -> energyStorage.capacity
        is IEnergyStorage -> energyStorage.maxEnergyStored.toLong()
        else -> 0
    }
    
    /**
     * Wrapper class that extends our EnergyStorage abstract class using Forge's IEnergyStorage.
     */
    class ForgeEnergyWrapper(
        private val forgeStorage: IEnergyStorage
    ) : EnergyStorage() {
        
        override val amount: Long get() = forgeStorage.energyStored.toLong()
        
        override val capacity: Long get() = forgeStorage.maxEnergyStored.toLong()
        
        fun simulateExtract(maxExtract: Long): Long = 
            forgeStorage.extractEnergy(min(maxExtract, Int.MAX_VALUE.toLong()).toInt(), true).toLong()
        
        fun simulateInsert(maxInsert: Long): Long = 
            forgeStorage.receiveEnergy(min(maxInsert, Int.MAX_VALUE.toLong()).toInt(), true).toLong()
        
        fun extract(maxExtract: Long, simulate: Boolean): Long = 
            forgeStorage.extractEnergy(min(maxExtract, Int.MAX_VALUE.toLong()).toInt(), simulate).toLong()
        
        fun insert(maxInsert: Long, simulate: Boolean): Long = 
            forgeStorage.receiveEnergy(min(maxInsert, Int.MAX_VALUE.toLong()).toInt(), simulate).toLong()
        
        fun getForgeStorage() = forgeStorage
    }
}